modules/gapi/src/backends/ie/giebackend.cpp

```diff
@@ -621,54 +725,54 @@ static void configureInputInfo(const IE::InputInfo::Ptr& ii, const cv::GMetaArg
+        copyFromIE(this_blob, out_mat);
+        ctx->out.post(ctx->output(i));
     }
+}
+
+struct PostListState {
+    PostListState(std::vector<std::vector<int>>&& cd)
+        : cached_dims(std::move(cd)) {
+        }
 
-    IECompiled                            &iec ;
-    std::shared_ptr<IECallContext>         ctx ;
+    std::vector<std::vector<int>> cached_dims;
+    size_t ndone = 0u;
 };
 
 // NB: This is a callback used by async infer
 // to post output list of blobs (cv::GArray<cv::GMat>).
-struct PostOutputsList {
-    // NB: Should be const to pass into SetCompletionCallback
-    void operator()() const {
-        for (auto i : ade::util::iota(ctx->uu.params.num_out)) {
-            std::vector<cv::Mat> &out_vec = ctx->outVecR<cv::Mat>(i);
-
-            IE::Blob::Ptr out_blob = iec.this_request.GetBlob(ctx->uu.params.output_names[i]);
+static void PostOutputsList(IECompiled                            &iec,
+                            std::shared_ptr<IECallContext>         ctx,
+                            size_t                                 position) {
+    // FIXME: Poor syncronization approach, just a global lock.
+    std::lock_guard<std::mutex>(ctx->m);
+    auto& state = ctx->state.get<PostListState>();
+    for (auto i : ade::util::iota(ctx->uu.params.num_out)) {
+        std::vector<cv::Mat> &out_vec = ctx->outVecR<cv::Mat>(i);
+
+        IE::Blob::Ptr out_blob = iec.this_request.GetBlob(ctx->uu.params.output_names[i]);
+
+        cv::Mat out_mat(state.cached_dims[i], toCV(out_blob->getTensorDesc().getPrecision()));
+        // FIXME: Avoid data copy. Not sure if it is possible though
+        copyFromIE(out_blob, out_mat);
+        out_vec[position] = std::move(out_mat);
+    }
 
-            cv::Mat out_mat(cached_dims[i], toCV(out_blob->getTensorDesc().getPrecision()));
-            // FIXME: Avoid data copy. Not sure if it is possible though
-            copyFromIE(out_blob, out_mat);
-            out_vec.push_back(std::move(out_mat));
-        }
-        // NB: Callbacks run synchronously yet, so the lock isn't necessary
-        auto&& out_vec_size = ctx->outVecR<cv::Mat>(0).size();
-        // NB: Now output vector is collected and can be posted to output
-        if (nrequests == out_vec_size) {
-            for (auto i : ade::util::iota(ctx->uu.params.num_out)) {
-                 ctx->out.post(ctx->output(i));
-            }
+    state.ndone++;
+    // Now output vector is collected, let's push it.
+    if (state.ndone == ctx->outVecR<cv::Mat>(0).size()) {
+        for (auto i : ade::util::iota(ctx->uu.params.num_out)) {
+            ctx->out.post(ctx->output(i));
         }
-
-        ctx->sync.release_and_notify();
     }
-
-    IECompiled                            &iec ;
-    std::shared_ptr<IECallContext>         ctx ;
-    std::vector< std::vector<int> >        cached_dims;
-    size_t                                 nrequests;
-};
+}
```

Comment:
This is probably too much. Can we focus on a single infer supporting `nireq` in scope of this review? So to make `infer-list` and `infer-2` continue working as usual?

Reply:
I hope it can be a task which is re-run IECompiled to process next ROI, thus ROi's will be processed consistently and synchronously within the single task scope.

Reply:
Implemented for `InferList`